guide/src/dev/index.md

```diff
@@ -0,0 +1,64 @@
+# Developer Guide
+
+pulldown-cmark uses a two-pass parsing strategy with a pull parser architecture to efficiently parse Markdown into HTML. This guide explains the internal workings of the library for developers who want to contribute or better understand how it works.
+
+## High-Level Architecture
+
+The parser operates in two main passes:
+
+1. **First Pass (Block Structure)**: The first pass scans the document and builds a tree structure representing block-level elements like paragraphs, lists, code blocks, etc. This establishes the hierarchical structure of the document.
+
+2. **Second Pass (Inline Processing)**: The second pass processes inline elements like emphasis, links and code spans within the blocks identified by the first pass. This is done in a streaming fashion as events are requested.
+
+The library uses a pull parser design, which means:
+
+- Instead of pushing events to a callback or building a complete AST, it provides an iterator interface that lets consumers pull events as needed
+- This allows for memory-efficient processing since only a small part of the document needs to be in memory at any time
+- It enables flexible transformation of the event stream before rendering
+
+Key components:
+
+- `Parser`: The main entry point that implements the Iterator trait for Events
+- `Tree`: A Vec-based data structure that holds the block structure 
+- `Event`: An enum representing the different Markdown elements
+- `HtmlWriter`: Renders the event stream as HTML
+
+## Performance Characteristics
+
+The parser is designed for high performance:
+
+- Memory usage is O(n) in the nesting depth of blocks, not document size
+- String handling uses copy-on-write semantics to avoid unnecessary allocations
+- SIMD optimizations are available for scanning text on x86_64
+- The pull parser design allows streaming without buffering the whole document
+
+## Extending the Parser
+
+The parser can be extended in several ways:
+
+- New syntax extensions can be added by implementing new scan functions
+- The event stream can be transformed using Iterator adaptors
+- Custom renderers can be built by consuming events
+- The HTML renderer can be customized through options
+
+## Directory Structure
```

Comment:
This might be a good place to mention the higher level directory structure too, not just the main crate under `pulldown-cmark/`.